core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -2412,7 +2423,235 @@ protected Scope visitUpdate(Update update, Optional<Scope> scope)
+            StatementAnalyzer analyzer = new StatementAnalyzer(
+                    analysis,
+                    metadata,
+                    sqlParser,
+                    groupProvider,
+                    new AllowAllAccessControl(),
+                    session,
+                    warningCollector,
+                    CorrelationSupport.ALLOWED);
+
+            Scope targetTableScope = analyzer.analyzeForUpdate(table, scope, UpdateKind.MERGE);
+
+            if (merge.getTargetAlias().isPresent()) {
+                String value = merge.getTargetAlias().get().getCanonicalValue();
+
+                RelationType relationType = targetTableScope.getRelationType();
+                RelationType descriptor = relationType.withAlias(value, null);
+
+                targetTableScope = createAndAssignScope(merge.getTargetAlias().get(), scope, descriptor);
+            }
+            Scope sourceTableScope = process(merge.getSource(), scope);
+            Scope joinScope = createAndAssignScope(merge, scope, targetTableScope.getRelationType().joinWith(sourceTableScope.getRelationType()));
+
+            TableSchema tableSchema = metadata.getTableSchema(session, targetTableHandle);
+
+            List<ColumnSchema> dataColumnSchemas = tableSchema.getColumns().stream()
+                    .filter(column -> !column.isHidden())
+                    .collect(toImmutableList());
+
+            Map<String, ColumnHandle> allColumnHandles = metadata.getColumnHandles(session, targetTableHandle);
+            ImmutableList.Builder<ColumnHandle> dataColumnHandlesBuilder = ImmutableList.builder();
+            ImmutableSet.Builder<String> dataColumnNamesBuilder = ImmutableSet.builder();
+            for (ColumnSchema columnSchema : dataColumnSchemas) {
+                String name = columnSchema.getName();
+                ColumnHandle handle = requireNonNull(allColumnHandles.get(name), "ColumnHandle for name is null");
+                dataColumnNamesBuilder.add(name);
+                dataColumnHandlesBuilder.add(handle);
+            }
+            List<ColumnHandle> dataColumnHandles = dataColumnHandlesBuilder.build();
+            Set<String> dataColumnNames = dataColumnNamesBuilder.build();
+
+            Map<String, Type> dataColumnTypes = dataColumnSchemas.stream().collect(toImmutableMap(ColumnSchema::getName, ColumnSchema::getType));
+
+            // Analyze all expressions in the Merge node
+
+            Expression mergePredicate = merge.getPredicate();
+            ExpressionAnalysis predicateAnalysis = analyzeExpression(mergePredicate, joinScope);
+            Type mergePredicateType = predicateAnalysis.getType(mergePredicate);
+            if (!typeCoercion.canCoerce(mergePredicateType, BOOLEAN)) {
+                throw semanticException(TYPE_MISMATCH, mergePredicate, "The MERGE predicate must evaluate to a boolean: actual type %s", mergePredicateType);
+            }
```

Comment:
You also need to record coercion if that is not BOOLEAN.

Reply:
Added.